feat: add `blas/base/wasm/srotm`

gururaj1512 opened the pull request:
Progresses #2039

## Description

> What is the purpose of this pull request?

This pull request:

-   This pull request adds `wasm` support for `srotm`.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Progresses #2039

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add `blas/base/wasm/srotm`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

gururaj1512 commented:
@kgryte @aman-095, when running tests, the arrays passed to srotm are not being updated. However, when running examples, they return as expected. Unable to identify the bug...

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/srotm/lib/routine.js:
```diff
@@ -0,0 +1,203 @@
+	Module.prototype.ndarray.call( this, N, p0.ptr, p0.stride, p0.offset, p1.ptr, p1.stride, p1.offset, p2.ptr ); // eslint-disable-line max-len
+
+	// If the arrays data had to be copied to module memory, copy the results to the provided corresponding output arrays...
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/srotm/lib/routine.js:
```diff
@@ -0,0 +1,203 @@
+	if ( p1.copy ) {
+		readDataView( N, this.view, p1.stride*p1.BYTES_PER_ELEMENT, p1.ptr, y, strideY, offsetY, true ); // eslint-disable-line max-len
+	}
+
```
```suggestion
```

kgryte commented:
> when running tests, the arrays passed to srotm are not being updated. However, when running examples, they return as expected. Unable to identify the bug...

@gururaj1512 The reason your tests are failing is that you are providing the wrong type of array for `param`. You need to provide a `Float32Array`, not a "generic" array. Wasm modules are typed, and, if you don't provide an array of the correct type, the offsets in module memory are off.

gururaj1512 commented:
Ohh 😅, Thanks.

gururaj1512 committed:
Update lib/node_modules/@stdlib/blas/base/wasm/srotm/lib/routine.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
Update lib/node_modules/@stdlib/blas/base/wasm/srotm/lib/routine.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
test: fix test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/wasm/srotm/docs/types/test.ts:
```diff
@@ -0,0 +1,614 @@
+	});
+	const mod = srotm.Module( mem );
+
+	mod.main( 10, 0, 1, 80, 1, '10'); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, true); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, false); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, null); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, undefined); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, []); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, {}); // $ExpectError
+	mod.main( 10, 0, 1, 80, 1, ( x: number ): number => x); // $ExpectError
```
```suggestion
	mod.main( 10, 0, 1, 80, 1, '10' ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, true ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, false ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, null ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, undefined ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, [] ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, {} ); // $ExpectError
	mod.main( 10, 0, 1, 80, 1, ( x: number ): number => x ); // $ExpectError
```

kgryte committed:
style: fix spacing

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update comments to be consistent

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM
